qutebrowser/browser/downloads.py

```diff
@@ -894,110 +894,196 @@ def _on_data_changed(self, idx, *, webengine):
             qtutils.ensure_valid(end_index)
         self.dataChanged.emit(start_index, end_index)
 
-    def _raise_no_download(self, count):
-        """Raise an exception that the download doesn't exist.
+    def _selective_dos(self, count, all_, indexset, criterion, dos,
```

Comment:
The problem is that, if I delete download `self[5]`, then immediately what was download `self[6]` will get pointed by index 5 instead. I.e. indexes change as we modify downloads. So what was `self[6]` will become `self[5]` if we delete `self[5]`.

This is why I think it's better to identify all downloads:
  1. Matching criteria.
  2. Not matching criteria.
  3. Not existing.

And put them all in their bags, all before we do anything to them, so that once we start doing things to them (e.g. deleting them) and causing their indexes to change, we will not have to figure out "hmm was the index of this download previously (say) 9?", because we have already put them in bags.

Then goes printing error messages. I am aggregating all errors for all downloads, so that I show them in one message. E.g. "Downloads 1,5-7,9,l-100 are not done yet!". Doing such error messages will require me to kinda aggregate all downloads first, and them summarize them.

Of course, I don't think it's impossible to do that with an iterator. But I just don't see how if I do it with an iterator, it will be simpler. 

Maybe I'm wrong. Afterall, this is the 1st time I use a lambda for anything useful. Maybe you know something that I don't. If that's the case, I'd appreciate if you show me how using a generator/iterator would help here.